### PR TITLE
717 - Validate json schema

### PR DESCRIPTION
**Pull request by brownbearsoftware**
**Validate against JSON schema when fetching applications**

When retrieving all JSON schemas for a user, test any that were created with an earlier version of the applications schema against the newest version of the application schema.  If they pass the validation, bump their version to the newest schema.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/ApplicationService.kt [hunk 1]
```diff
@@ -3,21 +3,40 @@ package uk.gov.justice.digital.hmpps.approvedpremisesapi.service
...
+        return application.apply { schemaUpToDate = false }
       }
+
+      application.schemaVersion = newestSchema
+      applicationRepository.save(application)
+    }
+
+    application.schemaUpToDate = true
+
+    return application
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/ApplicationService.kt:40]: I wonder if this belongs in the json Schema Service?

**Review by pezholio — APPROVED** [review]
LGTM 👍

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/JsonSchemaServiceTest.kt [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.factory.ApplicationEntityFactory
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.factory.ApplicationSchemaEntityFactory
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.factory.ProbationOfficerEntityFactory
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.ApplicationEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.ApplicationRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.ApplicationSchemaRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.service.JsonSchemaService
+import java.util.UUID
+
+class JsonSchemaServiceTest {
```
Comment (pezholio) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/JsonSchemaServiceTest.kt:17]: Was there a test for `validate` before? Would be good to have both

Comment (brownbearsoftware) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/JsonSchemaServiceTest.kt:17]: So this isn't actually called by anything other than `attemptSchemaUpgrade` - had meant to make it private hence the lack of tests.  Have made private now 👍

Comment (pezholio) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/JsonSchemaServiceTest.kt:17]: Ah, gotcha. I imagine it'll be used elsewhere at some point though right?

**Review by pezholio — APPROVED** [review]

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/JsonSchemaServiceTest.kt (hunk 2 above)
Comment (brownbearsoftware) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/JsonSchemaServiceTest.kt:17]: Yeah will need it during creation, will pop the unit tests in for it then

**Pull request merged**